File: broadcast.rb
require 'json'
require 'net/http'
require 'uri'
require 'taglib'
require 'net/imap'
require 'openai'
require 'time'

class Broadcast < Struct.new(:created_at, :events, :script_prompt, :script, :audio_file, :error)
  FILE_PATH = 'broadcast.json'

  def initialize(created_at: '', events: [], script_prompt: '', script: '', audio_file: '', error: '')
    super(created_at, events, script_prompt, script, audio_file, error)
  end

  def self.load
    return new unless File.exist?(FILE_PATH)

    data = JSON.parse(File.read(FILE_PATH))
    new(**data.transform_keys(&:to_sym))
  rescue JSON::ParserError
    new
  end

  def save
    File.write(FILE_PATH, JSON.pretty_generate(to_h))
    self
  end

  def to_h
    super.transform_keys(&:to_s)
  end
end

def ordinalize(number)
  abs_number = number.to_i.abs
  
  if (11..13).include?(abs_number % 100)
    "#{number}th"
  else
    case abs_number % 10
    when 1 then "#{number}st"
    when 2 then "#{number}nd"
    when 3 then "#{number}rd"
    else "#{number}th"
    end
  end
end

def get_weather_event
  uri = URI('https://api.openweathermap.org/data/3.0/onecall')
  uri.query = URI.encode_www_form({
    lat: ENV['LATITUDE'],
    lon: ENV['LONGITUDE'],
    appid: ENV['OPENWEATHERMAP_API_KEY'],
    units: 'imperial',
    exclude: 'minutely,hourly' # we only want daily forecasts
  })

  response = Net::HTTP.get_response(uri)
  weather_data = response.code == '200' ? JSON.parse(response.body) : nil

  current = weather_data["current"]
  today = weather_data["daily"][0]
  tomorrow = weather_data["daily"][1]
  right_now = Time.at(current['dt'])

  [
    'Today,',
    right_now.strftime("%A, %B the #{ordinalize(right_now.day)}"),
    "#{today['summary']}.",
    "Right now it's #{current['temp'].round} and feels like #{current['feels_like'].round}",
    "with a low of #{today['temp']['min'].round}",
    "with a high of #{today['temp']['max'].round}.",
    "Tomorrow, #{tomorrow['summary']}",
    "and a high of #{tomorrow['temp']['max'].round} and a heat index of #{tomorrow['feels_like']['day'].round}"
   ].join(' ')
end

def get_email_events
  imap = Net::IMAP.new(ENV['IMAP_HOST'], '993', true, nil, false)
  imap.authenticate('LOGIN', ENV['IMAP_USERNAME'], ENV['IMAP_PASSWORD'])
  imap.select('INBOX')

  search_terms = ["Shipped", "Out for Delivery", "Delivered", "on its way", "shipped", "on the way"]

  # IMAP querying is weird in ruby. You have to submit an array that
  # first contains your verbs, then you put in a field and search criteria
  # for each verb. So if I wanted to search a subject for "Amazon" OR "Walmart"
  # The array would be formatted as ["OR", "SUBJECT", "Amazon", "SUBJECT", "Walmart"]
  # weird, huh? The code below handles that.

  search_array = []
  # we need one OR for each pair of "FIELD" and "SEARCH TERMS", minus one because
  # we don't want a trailing "or"
  ((search_terms.count * 2) - 1).times { search_array << 'OR' }
  search_terms.each { |term| search_array += ['SUBJECT', term, 'SUBJECT', term.downcase] }
  email_seq_nos = imap.search(search_array)
  emails = if email_seq_nos == []
             []
           else
             imap.fetch(email_seq_nos, %w[ENVELOPE UID])
           end
  return [] if emails.count < 1

  puts "SHAZOOOM"

  emails.map do |email|
    puts email.attr['ENVELOPE'].date
    [
      "An email from #{email.attr['ENVELOPE'].from.first.name}",
      "with a subject of '#{email.attr['ENVELOPE'].subject.gsub(/\p{Emoji_Presentation}/, '').strip}'",
      "was received at #{email.attr['ENVELOPE'].date.strftime('%D %I:%M:%S %p')}"
    ].join(' ')
  end
end

def get_camera_events
  events = JSON.parse(`python3 get_wyze_events.py`)
  events.map do |event|
    event_time = DateTime.strptime(event['time'].to_s, '%Q').to_time.localtime
    event_text = "#{event['camera_name']} detected #{event['alarm_type']}"
    if event['tags'].count.positive?
      verb = event['alarm_type'].downcase == 'sound' ? 'heard' : 'saw'
      event_text = "#{event_text} and #{verb} #{event['tags'].map { |t| "a #{t}" }.join(' and ')}"
    end
    event_text   
  end
end

def get_thermostat_status
  status = JSON.parse(`python3 get_honeywell_status.py`)
  "The thermostat is set to #{status['mode']} and the indoor temperature is #{status['temperature']}"
end

def base_script_prompt
  <<-PROMPT.freeze
    In the style of a 1930's radio broadcaster, give a news update summarizing the below events.
    Do not include prompts, headers, or asterisks in the output.
    Do not read them all individually but group common events and summarize them.
    Do not include sound or music prompts. Mention that the broadcast is for the current time of #{Time.now.strftime('%I:00 %p')}
    The news update should be verbose and loquacious but please do not refer to yourself as either.
    The station name is 1.101 Cozy Castle Radio and your radio broadcaster name is Hotsy Totsy Harry Fitzgerald.
    At some point in the broadcast advertise for a ridiculous fictional product from the 1930's or tell a joke, do not do both.
    Give an introduction to the news report and a sign off.
    Here are the events:
  PROMPT
end

def generate_script(script_prompt)
  openai_client = OpenAI::Client.new(
    access_token: ENV['OPENAI_ACCESS_TOKEN'],
    organization_id: ENV['OPENAI_ORGANIZATION_ID']
  )

  messages = [{ role: 'user', content: script_prompt }]
  response = openai_client.chat(parameters: { model: 'gpt-5', messages: })
  response.dig('choices', 0, 'message', 'content')
end

def get_script_audio(text, filename)
  uri = URI("https://api.elevenlabs.io/v1/text-to-speech/#{ENV['ELEVENLABS_VOICE_ID']}/stream")
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true

  request = Net::HTTP::Post.new(uri)

  headers = {
    'xi-api-key' => ENV['ELEVENLABS_API_KEY'],
    'Content-Type' => 'application/json',
    'Accept' => 'audio/mpeg'
  }

  headers.each { |key, value| request[key] = value }
  request.body = { text: text, model_id: 'eleven_monolingual_v1' }.to_json

  File.open(filename, 'wb') do |file|
    http.request(request) do |response|
      raise StandardError, "Non-success status code while streaming #{response.code}" unless response.code == '200'

      response.read_body do |chunk|
        file.write(chunk)
      end
    end
  end

  filename
end

# Uses the library tool to read length in seconds from mp3 files
# brew install taglib
# env TAGLIB_DIR=/opt/homebrew/Cellar/taglib/2.1.1 gem install taglib-ruby --version '>= 2'
def audio_length_in_seconds(file_path)
  length = 0
  return length unless File.exist?(file_path)

  TagLib::MPEG::File.open(file_path) do |file|
    length = file.audio_properties.length_in_seconds
  end
  length
end

def mix_broadcast_audio(vocals_file, output_file)
    original_intro_outro_file = "radio_intro_outro.mp3"
    vocals_length = audio_length_in_seconds(vocals_file)
    intro_outro_length = audio_length_in_seconds(original_intro_outro_file)

    working_dir = 'work'
    intro_outro_file = "#{working_dir}/radio_intro_outro_resampled.mp3"
    padded_vocals_file = "#{working_dir}/padded_vocals.mp3".freeze
    faded_intro_outro_file = "#{working_dir}/faded_intro_outro.mp3"
    padded_outro_file = "#{working_dir}/padded_outro.mp3".freeze
    mixed_broadcast_file = "#{working_dir}/mixed_broadcast.wav".freeze
    compressed_broadcast_file = "#{working_dir}/compressed_broadcast.wav".freeze

    # create work directory and clear old file file
    system("mkdir work")
    system("rm #{output_file}")

    # resample file from Udio
    system("sox #{original_intro_outro_file} #{intro_outro_file} rate -h 44100")

    # pad vocals
    system("sox #{vocals_file} #{padded_vocals_file} pad 10@0")
    # fade intro outro
    system("sox #{intro_outro_file} #{faded_intro_outro_file} fade 5 25 20")
    # pad outro
    system("sox #{faded_intro_outro_file} #{padded_outro_file} pad #{vocals_length}@0")
    # mix files
    system("sox -M -v 0.3 #{faded_intro_outro_file} -v 1.2 #{padded_vocals_file} -v 0.4 #{padded_outro_file} #{mixed_broadcast_file}")
    # compress and fade in
    system("sox #{mixed_broadcast_file} #{compressed_broadcast_file} fade 5 compand 0.3,1 6:-70,-60,-20 -5 -90 0.2 norm -3")
    # rubocop:enable Layout/LineLength

    # wav to mono mp3
    system("sox #{compressed_broadcast_file} #{output_file} remix -")

    # cleanup everything
    system("rm -rf work")
    system("rm vocals_file.mp3")

    output_file
end

start_time = Time.now
puts "Starting Broadcast Generation... 🎙️"

record = Broadcast.load

record.created_at = Time.now.to_s

puts "Collecting Weather Data ☀️"
record.events << get_weather_event

puts "Collecting Email Events 📧"
record.events += get_email_events

puts "Collecting Camera Events 📹"
record.events += get_camera_events

puts "Collecting Thermostat Status 🌡️"
record.events << get_thermostat_status

record.script_prompt = "#{base_script_prompt} #{record.events.join("\n")}"

puts "Generating Script ✍️"
record.script = generate_script(record.script_prompt)

puts "Getting Vocals 🎤"
vocals_file = get_script_audio(record.script, "vocals_file.mp3")

puts "Mixing Audio 🎵"
record.audio_file = mix_broadcast_audio(vocals_file, "broadcast.mp3")

record.save

end_time = Time.now
elapsed_time = (end_time - start_time).round(2)
puts "You're done! 🎉 (Completed in #{elapsed_time} seconds)"

File: broadcast.py
import json
import os
import time
import requests
import imaplib
import email
from datetime import datetime, timedelta
from openai import OpenAI
from mutagen.mp3 import MP3
import subprocess
from pyhtcc import PyHTCC
from wyze_sdk import Client
from wyze_sdk.errors import WyzeApiError


class Broadcast:
    FILE_PATH = 'broadcast.json'

    def __init__(self, created_at='', events=None, script_prompt='', script='', audio_file='', error=''):
        if events is None:
            events = []
        self.created_at = created_at
        self.events = events
        self.script_prompt = script_prompt
        self.script = script
        self.audio_file = audio_file
        self.error = error

    @classmethod
    def load(cls):
        if not os.path.exists(cls.FILE_PATH):
            return cls()

        try:
            with open(cls.FILE_PATH, 'r') as f:
                data = json.load(f)
            return cls(**data)
        except (json.JSONDecodeError, FileNotFoundError):
            return cls()

    def save(self):
        with open(self.FILE_PATH, 'w') as f:
            json.dump(self.to_dict(), f, indent=2)
        return self

    def to_dict(self):
        return {
            'created_at': self.created_at,
            'events': self.events,
            'script_prompt': self.script_prompt,
            'script': self.script,
            'audio_file': self.audio_file,
            'error': self.error
        }


def ordinalize(number):
    abs_number = abs(int(number))

    if 11 <= abs_number % 100 <= 13:
        return f"{number}th"
    else:
        last_digit = abs_number % 10
        if last_digit == 1:
            return f"{number}st"
        elif last_digit == 2:
            return f"{number}nd"
        elif last_digit == 3:
            return f"{number}rd"
        else:
            return f"{number}th"


def get_weather_event():
    url = 'https://api.openweathermap.org/data/3.0/onecall'
    params = {
        'lat': os.environ['LATITUDE'],
        'lon': os.environ['LONGITUDE'],
        'appid': os.environ['OPENWEATHERMAP_API_KEY'],
        'units': 'imperial',
        'exclude': 'minutely,hourly'  # we only want daily forecasts
    }

    response = requests.get(url, params=params)
    if response.status_code != 200:
        return "Weather data unavailable"

    weather_data = response.json()
    current = weather_data["current"]
    today = weather_data["daily"][0]
    tomorrow = weather_data["daily"][1]
    right_now = datetime.fromtimestamp(current['dt'])

    parts = [
        'Today,',
        right_now.strftime(f"%A, %B the {ordinalize(right_now.day)}"),
        f"{today['summary']}.",
        f"Right now it's {round(current['temp'])} and feels like {round(current['feels_like'])}",
        f"with a low of {round(today['temp']['min'])}",
        f"with a high of {round(today['temp']['max'])}.",
        f"Tomorrow, {tomorrow['summary']}",
        f"and a high of {round(tomorrow['temp']['max'])} and a heat index of {round(tomorrow['feels_like']['day'])}"
    ]

    return ' '.join(parts)


def get_email_events():
    try:
        mail = imaplib.IMAP4_SSL(os.environ['IMAP_HOST'])
        mail.login(os.environ['IMAP_USERNAME'], os.environ['IMAP_PASSWORD'])
        mail.select('inbox')

        search_terms = ["Shipped", "Out for Delivery", "Delivered", "on its way", "shipped", "on the way"]

        # Search for each term individually and combine results
        # This avoids complex OR syntax issues in Python imaplib
        all_email_ids = set()

        for term in search_terms:
            try:
                # Search for the term (case-sensitive)
                status, messages = mail.search(None, 'SUBJECT', f'"{term}"')
                if status == 'OK' and messages[0]:
                    all_email_ids.update(messages[0].split())

                # Search for lowercase version if different
                if term != term.lower():
                    status, messages = mail.search(None, 'SUBJECT', f'"{term.lower()}"')
                    if status == 'OK' and messages[0]:
                        all_email_ids.update(messages[0].split())
            except:
                continue

        email_ids = list(all_email_ids)

        if not email_ids:
            mail.close()
            mail.logout()
            return []

        emails = []
        for email_id in email_ids:
            status, msg_data = mail.fetch(email_id, '(RFC822)')
            msg = email.message_from_bytes(msg_data[0][1])

            from_header = msg.get('From', '')
            subject = msg.get('Subject', '').encode('ascii', 'ignore').decode('ascii').strip()
            date_header = msg.get('Date', '')

            # Parse the date
            try:
                email_date = email.utils.parsedate_to_datetime(date_header)
                date_str = email_date.strftime('%m/%d/%y %I:%M:%S %p')
            except:
                date_str = date_header

            # Extract sender name (basic parsing)
            sender_name = from_header.split('<')[0].strip().strip('"') if '<' in from_header else from_header

            email_text = f"An email from {sender_name} with a subject of '{subject}' was received at {date_str}"
            emails.append(email_text)

        mail.close()
        mail.logout()
        return emails

    except Exception as e:
        print(f"Error getting email events: {e}")
        return []


def get_honeywell_status():
    try:
        p = PyHTCC(os.environ['HONEYWELL_USERNAME'], os.environ['HONEYWELL_PASSWORD'])
        zone = p.get_zone_by_name('THERMOSTAT')

        thermostat_data = {
            "temperature": zone.get_current_temperature_raw(),
            "mode": zone.get_system_mode().name
        }

        return f"The thermostat is set to {thermostat_data['mode']} and the indoor temperature is {thermostat_data['temperature']}"
    except Exception as e:
        print(f"Error getting thermostat status: {e}")
        return "Thermostat status unavailable"


def write_wyze_tokens_to_file(access_token, refresh_token):
    data = {
        "access_token": access_token,
        "refresh_token": refresh_token
    }
    with open("wyze_credentials.json", 'w+') as json_file:
        json.dump(data, json_file, indent=4)


def read_wyze_tokens_from_file():
    try:
        with open("wyze_credentials.json", 'r') as json_file:
            data = json.load(json_file)
        return data.get('access_token', ''), data.get('refresh_token', '')
    except FileNotFoundError:
        return '', ''


def populate_wyze_tokens():
    response = Client().login(
        email=os.environ['WYZE_EMAIL'],
        password=os.environ['WYZE_PASSWORD'],
        key_id=os.environ['WYZE_KEY_ID'],
        api_key=os.environ['WYZE_API_KEY']
    )
    write_wyze_tokens_to_file(response['access_token'], response['refresh_token'])


def get_camera_events():
    try:
        access_token, refresh_token = read_wyze_tokens_from_file()

        if not access_token:
            populate_wyze_tokens()
            access_token, refresh_token = read_wyze_tokens_from_file()

        client = Client(token=access_token, refresh_token=refresh_token)

        # Test current tokens
        try:
            client.cameras.list()
        except WyzeApiError as e:
            if str(e).startswith("The access token has expired"):
                resp = client.refresh_token()
                write_wyze_tokens_to_file(resp.data['data']['access_token'], resp.data['data']['refresh_token'])
                access_token, refresh_token = read_wyze_tokens_from_file()
                client = Client(token=access_token, refresh_token=refresh_token)

        # Pull all devices and build a map out of the mac address and the device name
        devices = client.devices_list()
        mac_map = {}
        for device in devices:
            mac_map[device.mac] = device.nickname

        twelve_hours_ago = datetime.now() - timedelta(hours=12)
        events = []

        for mac in mac_map.keys():
            wyze_events = client.events.list(device_ids=[mac], begin=twelve_hours_ago)
            for event in wyze_events:
                event_text = f"{mac_map[event.mac]} detected {event.alarm_type.description}"
                if event.tags:
                    verb = 'heard' if event.alarm_type.description.lower() == 'sound' else 'saw'
                    tags_text = ' and '.join([f"a {tag.description}" for tag in event.tags if tag is not None])
                    event_text = f"{event_text} and {verb} {tags_text}"
                events.append(event_text)

        return events

    except Exception as e:
        print(f"Error getting camera events: {e}")
        return []


def base_script_prompt():
    current_time = datetime.now().strftime('%I:00 %p')
    return f"""    In the style of a 1930's radio broadcaster, give a news update summarizing the below events.
    Do not include prompts, headers, or asterisks in the output.
    Do not read them all individually but group common events and summarize them.
    Do not include sound or music prompts. Mention that the broadcast is for the current time of {current_time}
    The news update should be verbose and loquacious but please do not refer to yourself as either.
    The station name is 1.101 Cozy Castle Radio and your radio broadcaster name is Hotsy Totsy Harry Fitzgerald.
    At some point in the broadcast advertise for a ridiculous fictional product from the 1930's or tell a joke, do not do both.
    Give an introduction to the news report and a sign off.
    Here are the events:"""


def generate_script(script_prompt):
    client = OpenAI(
        api_key=os.environ['OPENAI_ACCESS_TOKEN'],
        organization=os.environ['OPENAI_ORGANIZATION_ID']
    )

    response = client.chat.completions.create(
        model='gpt-4',
        messages=[{'role': 'user', 'content': script_prompt}]
    )

    return response.choices[0].message.content


def get_script_audio(text, filename):
    url = f"https://api.elevenlabs.io/v1/text-to-speech/{os.environ['ELEVENLABS_VOICE_ID']}/stream"

    headers = {
        'xi-api-key': os.environ['ELEVENLABS_API_KEY'],
        'Content-Type': 'application/json',
        'Accept': 'audio/mpeg'
    }

    data = {
        'text': text,
        'model_id': 'eleven_monolingual_v1'
    }

    response = requests.post(url, headers=headers, json=data, stream=True)

    if response.status_code != 200:
        raise Exception(f"Non-success status code while streaming {response.status_code}")

    with open(filename, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)

    return filename


def audio_length_in_seconds(file_path):
    if not os.path.exists(file_path):
        return 0

    try:
        audio = MP3(file_path)
        return int(audio.info.length)
    except:
        return 0


def mix_broadcast_audio(vocals_file, output_file):
    original_intro_outro_file = "radio_intro_outro.mp3"
    vocals_length = audio_length_in_seconds(vocals_file)
    intro_outro_length = audio_length_in_seconds(original_intro_outro_file)

    working_dir = 'work'
    intro_outro_file = f"{working_dir}/radio_intro_outro_resampled.mp3"
    padded_vocals_file = f"{working_dir}/padded_vocals.mp3"
    faded_intro_outro_file = f"{working_dir}/faded_intro_outro.mp3"
    padded_outro_file = f"{working_dir}/padded_outro.mp3"
    mixed_broadcast_file = f"{working_dir}/mixed_broadcast.wav"
    compressed_broadcast_file = f"{working_dir}/compressed_broadcast.wav"

    # create work directory and clear old file
    subprocess.run(["mkdir", "-p", "work"], check=False)
    subprocess.run(["rm", "-f", output_file], check=False)

    # resample file from Udio
    subprocess.run(["sox", original_intro_outro_file, intro_outro_file, "rate", "-h", "44100"], check=True)

    # pad vocals
    subprocess.run(["sox", vocals_file, padded_vocals_file, "pad", "10@0"], check=True)

    # fade intro outro
    subprocess.run(["sox", intro_outro_file, faded_intro_outro_file, "fade", "5", "25", "20"], check=True)

    # pad outro
    subprocess.run(["sox", faded_intro_outro_file, padded_outro_file, "pad", f"{vocals_length}@0"], check=True)

    # mix files
    subprocess.run([
        "sox", "-M",
        "-v", "0.3", faded_intro_outro_file,
        "-v", "1.2", padded_vocals_file,
        "-v", "0.4", padded_outro_file,
        mixed_broadcast_file
    ], check=True)

    # compress and fade in
    subprocess.run([
        "sox", mixed_broadcast_file, compressed_broadcast_file,
        "fade", "5", "compand", "0.3,1", "6:-70,-60,-20", "-5", "-90", "0.2", "norm", "-3"
    ], check=True)

    # wav to mono mp3
    subprocess.run(["sox", compressed_broadcast_file, output_file, "remix", "-"], check=True)

    # cleanup everything
    subprocess.run(["rm", "-rf", "work"], check=False)
    subprocess.run(["rm", "-f", "vocals_file.mp3"], check=False)

    return output_file


if __name__ == "__main__":
    start_time = time.time()
    print("Starting Broadcast Generation... 🎙️")

    record = Broadcast.load()
    record.created_at = datetime.now().isoformat()

    print("Collecting Weather Data ☀️")
    record.events.append(get_weather_event())

    print("Collecting Email Events 📧")
    record.events.extend(get_email_events())

    print("Collecting Camera Events 📹")
    record.events.extend(get_camera_events())

    print("Collecting Thermostat Status 🌡️")
    record.events.append(get_honeywell_status())

    record.script_prompt = f"{base_script_prompt()}\n{chr(10).join(record.events)}"

    print("Generating Script ✍️")
    record.script = generate_script(record.script_prompt)

    print("Getting Vocals 🎤")
    vocals_file = get_script_audio(record.script, "vocals_file.mp3")

    print("Mixing Audio 🎵")
    record.audio_file = mix_broadcast_audio(vocals_file, "broadcast.mp3")

    record.save()

    end_time = time.time()
    elapsed_time = round(end_time - start_time, 2)
    print(f"You're done! 🎉 (Completed in {elapsed_time} seconds)")